sklearn/linear_model/_base.py

```diff
@@ -287,11 +315,33 @@ def predict(self, X):
 
     def _set_intercept(self, X_offset, y_offset, X_scale):
         """Set the intercept_"""
+
+        xp, _ = get_namespace(X_offset, y_offset, X_scale)
+
         if self.fit_intercept:
             # We always want coef_.dtype=X.dtype. For instance, X.dtype can differ from
             # coef_.dtype if warm_start=True.
-            self.coef_ = np.divide(self.coef_, X_scale, dtype=X_scale.dtype)
-            self.intercept_ = y_offset - np.dot(X_offset, self.coef_.T)
+            coef_ = self.coef_ = xp.divide(
+                xp.astype(self.coef_, X_scale.dtype), X_scale
+            )
+
+            ravel = False
+            if self.coef_.ndim == 1:
+                coef_ = xp.reshape(coef_, (1, -1))
+                ravel = True
+
+            intercept_ = y_offset - (X_offset @ (coef_.T))
+
+            if ravel:
+                intercept_ = _asarray_with_order(
+                    intercept_, dtype=intercept_.dtype, order="C", copy=None, xp=xp
+                )
```

Comment:
> I don't understand why this would ever be needed

most likely some tests were breaking without that, I'll check again.

Reply:
numpy-based tests keep passing without this instruction so I suspect the issue was with other libraries that print `array(1)` rather than `1` in some tests, I presume it can also be dealt with by using `float` rather than `reshape` + `getitem`. Will fix.